libgadget/lenstools.c

```diff
@@ -0,0 +1,412 @@
+#include "lenstools.h"
+#include <mpi.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <fftw3.h> 
+
+#ifdef USE_CFITSIO
+#include "fitsio.h"
+#endif
+
+#include <string.h>
+#include "partmanager.h"
+#include "cosmology.h"
+#include "physconst.h"
+
+void linspace(double start, double stop, int num, double *result) {
+    double step = (stop - start) / (num - 1);
+    for (int i = 0; i < num; i++) {
+        result[i] = start + i * step;
+    }
+}
+
+double ***allocate_3d_array(int nx, int ny, int nz) {
+    // Allocate memory for the 3D array
+    double ***array = malloc(nx * sizeof(double **));
```

Comment:
In MP-Gadget we use mymalloc/myfree which has a code-managed heap to avoid memory fragmentation. The restriction is that memory has to be de-allocated in the reverse order that it is allocated. So you write something like:

array = mymalloc("3ds", nx*sizeof(double **));
array[i] = mymalloc("2ds", ny * sizeof(double *));
array[i][j] = ...
(stuff)
myfree(array[i][j]);
myfree(array[i]);
myfree(array);

Reply:
I rewrite the function as:

double ***allocate_3d_array_new(int nx, int ny, int nz) {
    // Allocate memory for the 3D array
    double ***array = mymalloc("3ds", nx * sizeof(double **));
    if (array == NULL) {
        fprintf(stderr, "Memory allocation failed\n");
        exit(EXIT_FAILURE);
    }

    for (int i = 0; i < nx; i++) {
        array[i] = mymalloc("2ds", ny * sizeof(double *));
        if (array[i] == NULL) {
            fprintf(stderr, "Memory allocation failed for layer %d\n", i);
            exit(EXIT_FAILURE);
        }
        for (int j = 0; j < ny; j++) {
            array[i][j] = mymalloc("1ds",nz * sizeof(double));
            if (array[i][j] == NULL) {
                fprintf(stderr, "Memory allocation failed for layer %d, row %d\n", i, j);
                exit(EXIT_FAILURE);
            }
            // Initialize elements (optional)
            for (int k = 0; k < nz; k++) {
                array[i][j][k] = 0.0; // Example initialization
            }
        }
    }

    return array;
}

But got error messages like:

1ds                  | B | 000000000000 000000000004 | lenstools.c:70
[ 000003.86 ] Task 1:  1ds                  | B | 000000000000 000000000004 | lenstools.c:70
[ 000003.86 ] Task 0:  1ds                  | B | 000000000000 000000000004 | lenstools.c:70
[ 000003.86 ] Task 1:  1ds                  | B | 000000000000 000000000004 | lenstools.c:70
[ 000003.86 ] Task 1:  1ds                  | B | 000000000000 000000000004 | lenstools.c:70
[ 000003.86 ] Task 0:  1ds                  | B | 000000000000 000000000004 | lenstools.c:70
[ 000003.86 ] Task 0:  1ds                  | B | 000000000000 000000000004 | lenstools.c:70
[ 000003.86 ] Task 0: [ 000003.86 ] Task 1:  1ds                  | B | 000000000000 000000000004 | lenstools.c:70
[ 000003.86 ] Task 1:  1ds                  | B | 000000000000 000000000004 | lenstools.c:70
[ 000003.86 ] Task 0:  1ds                  | B | 000000000000 000000000004 | lenstools.c:70
 1ds                  | B | 000000000000 000000000004 | lenstools.c:70
[ 000003.86 ] Task 1:  TopLeaves            | T | 000000000000 000000000004 | domain.c:198
[ 000003.86 ] Task 0:  1ds                  | B | 000000000000 000000000004 | lenstools.c:70
[ 000003.86 ] Task 1:  TopNodes             | T | 000000000000 000000000004 | domain.c:196
[ 000003.86 ] Task 1: [ 000003.86 ] Task 0:  1ds                  | B | 000000000000 000000000004 | lenstools.c:70
[ 000003.86 ] Task 0:  Tasks                | T | 000000000000 000000000004 | domain.c:374
[ 000003.86 ] Task 1:  Mesh2Task            | T | 000000000001 000000000004 | petapm.c:119
 1ds                  | B | 000000000000 000000000004 | lenstools.c:70
[ 000003.86 ] Task 0:  1ds                  | B | 000000000000 000000000004 | lenstools.c:70
[ 000003.86 ] Task 0: [ 000003.86 ] ERROR: Task 1: Not enough memory for 1ds 4096 bytes
 1ds                  | B | 000000000000 000000000004 | lenstools.c:70
[ 000003.86 ] Task 0:  1ds                  | B | 000000000000 000000000004 | lenstools.c:70

The way I call mymalloc might not be right?

Reply:
Maybe try doing a single allocation and setting each element to an offset within that allocation?
            array[0][0] = mymalloc("1ds",nx * ny * nz * sizeof(double));
for(i, j)
array[i][j] = array + i * nx *ny + ny * j;
(or something like that: I didn't work out the actual offsets here).

Reply:
I have rewritten those allocation functions and use mymalloc and myfree! (and access 2D and 3D arrays in a 1D style)

Reply:
Nice! I will take a look.